syntax = "proto3";

package ticket;

// The train ticketing service definition.
service TrainService {
	// API1
	rpc PurchaseTicket (TicketRequest) returns (TicketReceipt);

	// API2
	rpc GetReceiptDetails (ReceiptDetailsRequest) returns (TicketReceipt);
	
	// API3
	rpc ViewUsersBySection (SectionRequest) returns (UsersBySectionResponse);

	// API4
	rpc RemoveUser (RemoveUserRequest) returns (RemoveUserResponse);

	// API5
	rpc ModifyUserSeat (ModifySeatRequest) returns (ModifySeatResponse);
}

message TicketRequest{
	string fromTRQ = 1;
	string to = 2;
	User user = 3;
	float price_paid = 4;
}

message ReceiptDetailsRequest{
	string receipt_id = 1;
}

message SectionRequest{
	string section = 1;
}

message RemoveUserRequest {
	string email = 1;
}

message ModifySeatRequest {
	string email = 1;
	string new_seat = 2;
}

message User {
	string first_name = 1;
	string last_name = 2;
	string email = 3;
	float price_paid = 4;
}

message TicketReceipt {
	string receipt_id = 1;
	string fromTRP = 2;
	string to = 3;
	User user = 4;
	float price_paid = 5;
	string seat = 6;
}


message UsersBySectionResponse{
	repeated UserSeat user_seat = 1;
}

message UserSeat {
	string user_id = 1;
	string seat = 2;
}

message RemoveUserResponse{
	bool success = 1;
}

message ModifySeatResponse{
	bool success = 1;
}

